source/extensions/filters/udp/udp_proxy/router/router_impl.cc

```diff
@@ -0,0 +1,83 @@
+#include "source/extensions/filters/udp/udp_proxy/router/router_impl.h"
+
+#include "envoy/extensions/filters/udp/udp_proxy/v3/route.pb.h"
+#include "envoy/extensions/filters/udp/udp_proxy/v3/route.pb.validate.h"
+
+#include "source/common/common/empty_string.h"
+#include "source/common/network/matching/data_impl.h"
+#include "source/common/network/matching/inputs.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace UdpFilters {
+namespace UdpProxy {
+namespace Router {
+
+Matcher::ActionFactoryCb RouteMatchActionFactory::createActionFactoryCb(
+    const Protobuf::Message& config, RouteActionContext& context,
+    ProtobufMessage::ValidationVisitor& validation_visitor) {
+  const auto& route_config = MessageUtil::downcastAndValidate<
+      const envoy::extensions::filters::udp::udp_proxy::v3::Route&>(config, validation_visitor);
+  const auto& cluster = route_config.cluster();
+
+  // Emplace cluster names to context to get all cluster names
+  context.cluster_name_.emplace(cluster);
+
+  return [cluster]() { return std::make_unique<RouteMatchAction>(cluster); };
+}
+
+REGISTER_FACTORY(RouteMatchActionFactory, Matcher::ActionFactory<RouteActionContext>);
+
+absl::Status RouteActionValidationVisitor::performDataInputValidation(
+    const Matcher::DataInputFactory<Network::UdpMatchingData>&, absl::string_view) {
+  return absl::OkStatus();
+}
+
+RouterImpl::RouterImpl(const envoy::extensions::filters::udp::udp_proxy::v3::UdpProxyConfig& config,
+                       Server::Configuration::ServerFactoryContext& factory_context) {
+  if (config.has_cluster()) {
+    cluster_ = config.cluster();
```

Comment:
I think it would be a bit cleaner to just synthesize matcher config for this case. Then you can use unified code for the rest of it? (Basically it makes the deprecated path cheaper to maintain and understand as the only special case is here.)

Reply:
I am sorry but I cannot catch your idea here. Do you mention we can move `cluster_` out of `Router`?

Reply:
Ok, I catch your thought finally, I will submit the commit soon.